compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEUtils.cpp

```diff
@@ -107,6 +107,26 @@ FailureOr<unsigned> getTilingScaleFactor(Type elemType) {
   return 64 / bitWidth;
 }
 
+/// Utility to check if a generic op is an elementwise op and if its producer is
+/// a matmul-like op.
+bool isMatmulElementwiseFusion(linalg::LinalgOp linalgOp) {
+  if (!isElementwise(linalgOp)) return false;
+  // Check if any of the defining op is a matmul-like op. To simplify the
+  // problem, currently only check if it is a contraction op.
```

Comment:
Add a `TODO` before the comment.

I'm not sure if there'd be cases where any other contraction op needs to supported for the elementwise fusion.

But if there'll never be any then yes this needs to be made stricter.

Just adding here the utility which you may use to update this later in that case :-
```
/// Utility to match iterator type and indexing map for a linalg.generic that
/// is basically implementing a matmul with 6D input/output operands.
static bool match6DLinalgGenericMatmul(linalg::LinalgOp linalgOp) {
  // Check iterator types.
  SmallVector<utils::IteratorType> matmulIteratorTypes = {
      utils::IteratorType::parallel,  utils::IteratorType::parallel,
      utils::IteratorType::reduction, utils::IteratorType::parallel,
      utils::IteratorType::parallel,  utils::IteratorType::reduction,
      utils::IteratorType::parallel,  utils::IteratorType::parallel,
      utils::IteratorType::reduction};
  SmallVector<utils::IteratorType> opIteratorTypes =
      linalgOp.getIteratorTypesArray();
  if (matmulIteratorTypes != opIteratorTypes) {
    return false;
  }
  // Check indexing maps.
  ArrayAttr indexingMaps = linalgOp.getIndexingMaps();
  if (indexingMaps.size() != 3) return false;

  AffineMap map0 = cast<AffineMapAttr>(indexingMaps[0]).getValue();
  AffineMap map1 = cast<AffineMapAttr>(indexingMaps[1]).getValue();
  AffineMap map2 = cast<AffineMapAttr>(indexingMaps[2]).getValue();

  if (map0.getNumResults() != 6 || map1.getNumResults() != 6 ||
      map2.getNumResults() != 6 || map0.getNumInputs() != 9 ||
      map1.getNumInputs() != 9 || map2.getNumInputs() != 9) {
    return false;
  }

  // Since this is invoked after the final pack-and-transpose we need to
  // extract dimensions for M*K*k*m*m0*k0 x K*N*n*k*k0*n0 -> M*N*n*m*m0*n0.
  AffineExpr M = map2.getResult(0);
  AffineExpr N = map2.getResult(1);
  AffineExpr n = map2.getResult(2);
  AffineExpr m = map2.getResult(3);
  AffineExpr m0 = map2.getResult(4);
  AffineExpr n0 = map2.getResult(5);
  AffineExpr K = map0.getResult(1);
  AffineExpr k = map0.getResult(2);
  AffineExpr k0 = map0.getResult(5);

  auto *context = indexingMaps.getContext();
  auto mapA =
      AffineMapAttr::get(AffineMap::get(9, 0, {M, K, k, m, m0, k0}, context));
  auto mapB =
      AffineMapAttr::get(AffineMap::get(9, 0, {K, N, n, k, k0, n0}, context));
  auto mapC =
      AffineMapAttr::get(AffineMap::get(9, 0, {M, N, n, m, m0, n0}, context));
  auto maps = ArrayAttr::get(context, {mapA, mapB, mapC});
  return indexingMaps == maps;
}
```

Reply:
My original intention was to make it a general utility function and to be used in several passes, but I ended up deleted other uses. I'm not sure if it's a good idea to constrain the matmul to be 6D input and output. Maybe I can make a stricter matcher right now, and generalize it later if needed. What do you think?

Reply:
I moved some utility functions from `AMDAIELowerToUKernels.cpp` here as common utilities, so that we can apply these to match all 2D/4D/6D generic form matmuls.

Reply:
> generalize it later if needed. What do you think?

Sounds perfect!

> I moved some utility functions from AMDAIELowerToUKernels.cpp here as common utilities

Thanks for this @yzhang93 !